src/dataconnect/schemaMigration.ts

```diff
@@ -43,60 +46,97 @@ export async function migrateSchema(args: {
   allowNonInteractiveMigration: boolean;
   validateOnly: boolean;
 }): Promise<Diff[]> {
-  const { schema, validateOnly } = args;
+  const { options, schema, allowNonInteractiveMigration, validateOnly } = args;
 
   const databaseId = schema.primaryDatasource.postgresql?.database;
   if (!databaseId) {
     throw new FirebaseError(
       "Schema is missing primaryDatasource.postgresql?.database, cannot migrate",
     );
   }
-  const instanceId = schema.primaryDatasource.postgresql?.cloudSql.instance.split("/").pop();
-  if (!instanceId) {
-    throw new FirebaseError(`tried to migrate schema but ${instanceId} was undefined`);
+  const instanceName = schema.primaryDatasource.postgresql?.cloudSql.instance;
+  if (!instanceName) {
+    throw new FirebaseError(
+      "tried to migrate schema but instance name was not provided in dataconnect.yaml",
+    );
   }
+  const instanceId = instanceName?.split("/").pop();
+
+  const serviceName = schema.name.replace(`/schemas/${SCHEMA_ID}`, "");
   try {
-    const serviceName = schema.name.replace(`/schemas/${SCHEMA_ID}`, "");
-    await ensureServiceIsConnectedToCloudSql(serviceName);
+    await ensureServiceIsConnectedToCloudSql(serviceName, instanceName, databaseId);
     await upsertSchema(schema, validateOnly);
     logger.debug(`Database schema was up to date for ${instanceId}:${databaseId}`);
-    return [];
   } catch (err: any) {
-    const incompatible = getIncompatibleSchemaError(err);
-    if (!incompatible) {
+    const incompatible = errors.getIncompatibleSchemaError(err);
+    const invalidConnectors = errors.getInvalidConnectors(err);
+    if (!incompatible && !invalidConnectors.length) {
       // If we got a different type of error, throw it
       throw err;
     }
-    // Try to migrate schema
-    const diffs = await handleIncompatibleSchemaError({
-      ...args,
-      incompatibleSchemaError: incompatible,
-      instanceId,
-      databaseId,
-    });
+    const shouldDeleteInvalidConnectors = await promptForInvalidConnectorError(
+      options,
+      invalidConnectors,
+      validateOnly,
+    );
+    if (!shouldDeleteInvalidConnectors && invalidConnectors.length) {
+      const cmd = suggestedCommand(serviceName, invalidConnectors);
+      throw new FirebaseError(
+        `Command aborted. Try deploying compatible connectors first with ${clc.bold(cmd)}`,
+      );
+    }
+    const migrationMode = incompatible
+      ? await promptForSchemaMigration(
+          options,
+          databaseId,
+          incompatible,
+          allowNonInteractiveMigration,
+        )
+      : "none";
+    // First, error out if we aren't making all changes
+    if (migrationMode === "none" && incompatible) {
```

Comment:
No AI needed. 

Note that since backend checks for exact match of SQL schema. 

`migrationMode === "safe"` is in an awkward unusable state, since it will fail below:
```
    // Then, try to upsert schema again. If there still is an error, just throw it now
    await upsertSchema(schema, validateOnly);
```

Rosalyn's intern project "incompatible" mode would address this hole.

Ideas to improve it before there, we could suggest developers to use `deploy --force` to update FDC schema ignoring schema validation error. `schemaValidation=NONE`

Definitely out of scope for this PR. Post-IO polish idea.

Reply:
[For follow up]

non-force `firebase deploy` should block on schema migration being fully done. This following prompt makes sense
```
    const choices = err.destructive
      ? [
          { name: "Execute all changes (including destructive changes)", value: "all" },
          // This set schemaValidation=NONE in the second `upsertSchema` 
          { name: "Skip any destructive changes", value: "safe" }, 
          { name: "Abort changes", value: "abort" }, // non-interactive mode behavior
        ]
      : [
          { name: "Execute changes", value: "safe" },
          { name: "Abort changes", value: "abort" }, // non-interactive mode behavior
        ];
```

Interactive `deploy --force` prompts as well, but should provide more options to force deploy ignoring schema diffs.

```
    const choices = err.destructive
      ? [
          { name: "Execute all changes (including destructive changes)", value: "all" },
          // This set schemaValidation=NONE in the second `upsertSchema` 
          { name: "Skip any destructive changes", value: "safe" }, 
          // This set schemaValidation=NONE in the second `upsertSchema` 
          { name: "Skip all changes", value: "skip" }, // non-interactive mode behavior
          { name: "Abort changes", value: "abort" },
        ]
      : [
          { name: "Execute changes", value: "safe" },
          // This set schemaValidation=NONE in the second `upsertSchema` 
          { name: "Skip all changes", value: "skip" }, // non-interactive mode behavior
          { name: "Abort changes", value: "abort" },
        ];
```